integration/src/main/java/scripts/testscripts/RemoveUser.java

```diff
@@ -0,0 +1,212 @@
+    // Validate that setup ran correctly and users have appropriate resource access.
+    ResourceMaker.retrieveBucketFile(sharedBucketName, projectId, sharedResourceUser);
+    ResourceMaker.retrieveBucketFile(privateBucketName, projectId, privateResourceUser);
+
+    // Remove WRITER role from sharedResourceUser. This is their only role, so they are no longer
+    // a member of this workspace.
+    ownerWorkspaceApi.removeRole(getWorkspaceId(), IamRole.WRITER, sharedResourceUser.userEmail);
+
+    // Validate that sharedResourceUser can no longer read resources in the workspace.
+    // This requires syncing google groups, so there is often a delay that we need to wait for.
+    ClientTestUtils.runWithRetryOnException(
+        () -> assertUserCannotReadBucket(sharedBucketName, sharedResourceUser));
+
+    // privateResource user can still read the shared bucket.
+    ResourceMaker.retrieveBucketFile(sharedBucketName, projectId, privateResourceUser);
+
+    // Remove READER role from privateResourceUser. They are also a writer, so they should not lose
+    // access to workspace resources because of this.
+    ownerWorkspaceApi.removeRole(getWorkspaceId(), IamRole.READER, privateResourceUser.userEmail);
+
+    // Validate privateResourceWriter still has access to all resources.
+    ResourceMaker.retrieveBucketFile(sharedBucketName, projectId, privateResourceUser);
+    ResourceMaker.retrieveBucketFile(privateBucketName, projectId, privateResourceUser);
+    ResourceMaker.readPopulatedBigQueryTable(privateDataset, privateResourceUser, projectId);
+    assertTrue(ResourceMaker.userHasProxyAccess(privateNotebook, privateResourceUser, projectId));
+
+    // Remove WRITER role from privateResourceUser. This is their last role, so they are no longer
+    // a member of this workspace.
+    ownerWorkspaceApi.removeRole(getWorkspaceId(), IamRole.WRITER, privateResourceUser.userEmail);
+
+    // Validate privateResourceWriter no longer has access to any private resources.
+    ClientTestUtils.runWithRetryOnException(
+        () -> assertUserCannotReadBucket(sharedBucketName, privateResourceUser));
+    ClientTestUtils.runWithRetryOnException(
+        () -> assertUserCannotReadBucket(privateBucketName, privateResourceUser));
+    ClientTestUtils.runWithRetryOnException(
+        () -> assertUserCannotReadDataset(privateDataset, privateResourceUser));
+    ClientTestUtils.runWithRetryOnException(
+        () -> assertUserCannotAccessNotebook(privateNotebook, privateResourceUser));
+  }
+
+  /**
+   * An assertion that the given user cannot read from the given bucket. This is pulled into a
+   * separate function to make retrying simpler.
+   */
+  private void assertUserCannotReadBucket(String bucketName, TestUserSpecification testUser) {
+    try {
+      ResourceMaker.retrieveBucketFile(bucketName, projectId, testUser);
+      // If nothing is thrown, that's bad! The user actually can read the bucket.
+      throw new RuntimeException(String
```

Comment:
It does retry that exception, which is what we want since access hasn't been revoked yet. The "happy path" is getting a 403 storage exception in this case.

An alternative here is writing a `runWithRetryOnNoException` method - either way, these are looking to validate that the right exception is being thrown.

Reply:
Maybe indicate in the comment that this is only designed to work with that one retry function.